echo-pubsub-google/src/main/java/com/netflix/spinnaker/echo/pubsub/google/GooglePubsubNotificationEventListener.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.pubsub.google;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.echo.api.Notification;
+import com.netflix.spinnaker.echo.api.Notification.Type;
+import com.netflix.spinnaker.echo.config.GooglePubsubProperties.Content;
+import com.netflix.spinnaker.echo.controller.EchoResponse;
+import com.netflix.spinnaker.echo.controller.EchoResponse.Void;
+import com.netflix.spinnaker.echo.model.Event;
+import com.netflix.spinnaker.echo.model.pubsub.PubsubSystem;
+import com.netflix.spinnaker.echo.notification.AbstractEventNotificationAgent;
+import com.netflix.spinnaker.echo.notification.NotificationService;
+import com.netflix.spinnaker.echo.pubsub.PubsubPublishers;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+@ConditionalOnExpression("${pubsub.enabled:false} && ${pubsub.google.enabled:false}")
+public class GooglePubsubNotificationEventListener extends AbstractEventNotificationAgent implements
+  NotificationService {
+
+  private static Type TYPE = Type.PUBSUB;
+
+  @Autowired
+  private PubsubPublishers publishers;
+
+  @Autowired
+  ObjectMapper mapper;
+
+  /**
+   * Handles calls caught and processed by the superclass from the HistoryController invocations.
+   */
+  @Override
+  public void sendNotifications(Map notification, String application, Event event, Map config,
+    String status) {
+    publishers.publishersMatchingType(PubsubSystem.GOOGLE)
+      .stream()
+      .map(p -> (GooglePubsubPublisher) p)
+      .filter(p -> p.getContent() == Content.NOTIFICATIONS)
+      .filter(p -> notification.containsKey("topic") && notification.get("topic").toString()
+        .equalsIgnoreCase(p.getTopicName()))
+      .forEach(p -> p.publishEvent(event));
+  }
+
+  /**
+   * Handles explicit calls from the NotificationsController - generally only used with Manual
+   * Judgement stage from Orca.
+   */
+  @Override
+  public EchoResponse.Void handle(Notification notification) {
+    if (log.isDebugEnabled()) {
+      try {
+        log.debug("Notification received: " + mapper.writerWithDefaultPrettyPrinter()
+          .writeValueAsString(notification));
+      } catch (JsonProcessingException jpe) {
+        log.warn("Error parsing notification", jpe);
```

Comment:
Will this cause trouble later in this method? I'm wondering if this is validation we need to be doing in all cases & report on it.

Reply:
I don't ever expect this to be called, but because it's a checked exception, I have to try/catch it.